### PR TITLE
React integration

### PR DESCRIPTION
**Pull request by solomon-gumball**
- [x] Build the feature
- [x] Write all the tests

This PR allows the user to use `Navigator.js` to

Add REACT elements as children of a BACKBONE view
Add REACT elements as children of the root DOM element
Add REACT elements and children of REACT elements

Adds removal processes for both BACKBONE and REACT elements on state change.

Example:

```
this.stateViewMap
    .mapState('myapp')
    .toView(MyReactComponent)
    .withArguments({
        someProp: 'someValue'
    })
    .inside('body');
```
## How to test:

I created a test repo here: https://github.com/redwoodfavorite/navigator-react-test

To run the app, clone the repo and this repo.  Then run `npm link` in this repo.  Next go to the demo repo and npm install and run `npm link navigator-js`.  Now just run `npm run dev` to start the app on `localhost:8080`

### CODE REVIEW

File: public/js/navigator-js/integration/BackboneRecipe.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+var BackboneRecipe = {
```
Comment (bassettsj) [inline public/js/navigator-js/integration/BackboneRecipe.js:1]: @redwoodfavorite es6?

File: webpack.config.js [hunk 2]
```diff
@@ -18,12 +18,9 @@ module.exports = {
...
     module: {
         loaders: [{
-            test: /(\.jsx?)$/,
+            test: /\.js/,
             exclude: /(bower_components|node_modules)/,
-            loader: 'babel',
-            query: {
-                cacheDirectory: true
-            }
+            loader: 'imports?this=>window'
```
Comment (bassettsj) [inline webpack.config.js:23]: We should fix this...

File: public/js/navigator-js/integration/StateViewMap.js [hunk 3]
```diff
@@ -43,33 +45,90 @@ this.navigatorjs.integration = this.navigatorjs.integration || {};
...
+				} else {
+					if (recipe.isMounted()) {
+						_removeViewElementFromDOM(recipe);
+					}
 				}
 			}
 		}
 	}
 
+	function _removeViewElementFromDOM(recipe) {
```
Comment (elimak) [inline public/js/navigator-js/integration/StateViewMap.js]: We are currently not removing the views that were created, but we are hiding them until they transitionIn again. 
Based on the performances benchmark: 'Creation' is 5 to 30 times slower than 'Update'
If we decide to throw away our existing views and create them over and over, it will slow down significantly the pages loading time.

File: karma.conf.js [hunk 4]
```diff
@@ -9,7 +9,7 @@ const webpackConfig = {
             include: path.resolve('spec/'),
             exclude: [
                 path.resolve('src/'),
-                path.resolve('node_modules/')
+                path.resolve('node_modules/'),
```
Comment (gregegan) [inline karma.conf.js]: comma?

File: public/js/navigator-js/integration/ReactRecipe.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  _type: 'REACT',
+
+  // Returns a fake proxy element due to the fact that
+  // we do not use these transition functions within
+  // our react components
+
+  getViewInstance: function getViewInstance() {
+    return {
+      navigatorBehaviors: ['IHasStateTransition'],
+      transitionIn: function(cb) { cb() },
```
Comment (elimak) [inline public/js/navigator-js/integration/ReactRecipe.js]: I am not sure I get this. 
The concept behind transitionIn/Out is to allow for timed transition - 
transitionIn(cb) {
// view does some animation or wait for some data or anything timed....
// then transitionIn -> callback ===> trigger transitionOut for previous view instance 
cb();
}

By making these proxies are we fully losing this feature?

File: public/js/navigator-js/integration/StateViewMap.js (hunk 3 above)
Comment (elimak) [inline public/js/navigator-js/integration/StateViewMap.js]: Correct me if I am wrong - removing the view here only affect the DOM and does not destroy the js class. Which should be less expensive that disposing the full view. Right? 
Anyways, could we benchmark this in your demo? I would like to see the difference in the rendering time when removing / adding html node vs. display:none / ''

File: public/js/navigator-js/integration/ReactRecipe.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+  // Save reference to our react element instead of
+  // view instance like backbone recipe does
+
+  initialize: function initialize() {
+    var params = this._viewArguments;
+
+    var props = Object.assign(
+      {
+        ref: function(c) {
```
Comment (elimak) [inline public/js/navigator-js/integration/ReactRecipe.js:63]: what is `ref`?

File: public/js/navigator-js/integration/StateViewMap.js (hunk 3 above)
Comment (solomon-gumball) [inline public/js/navigator-js/integration/StateViewMap.js]: Yes, removing the element from the DOM will keep the view, the listeners and the element exactly as they are, nothing is garbage collected here.

http://stackoverflow.com/questions/12528049/if-a-dom-element-is-removed-are-its-listeners-also-removed-from-memory

You're right that toggling the `display` property of a DOM element is a faster way to hide an element than removing and adding the child node to the DOM.

https://jsperf.com/removing-an-element-from-the-screen/3

But for me this change is better from an abstraction standpoint.  The styles of the element should be determined by the view itself and it's presence in the document should be handled by the router.   In my opinion, when I run a `querySelector` on the page I should not be able to find all the elements that were hidden by the router.

File: public/js/navigator-js/integration/ReactRecipe.js (hunk 5 above)
Comment (solomon-gumball) [inline public/js/navigator-js/integration/ReactRecipe.js]: Yes we are.  I really wanted to make `transitionIn` and `transitionOut` work with React but there are some complications that would require a more involved refactor to Navigator due to the fact that react component references are not available immediately after `render` is called.

My proposal is that we start with this and address this issue in a follow up PR.

File: public/js/navigator-js/integration/ReactRecipe.js (hunk 6 above)
Comment (solomon-gumball) [inline public/js/navigator-js/integration/ReactRecipe.js:63]: `ref` is a function that is passed the reference to the react component instance when it changes.  Similar to what would be returned from a `new Backbone.View()`

File: public/js/navigator-js/integration/BackboneRecipe.js (hunk 1 above)
Comment (solomon-gumball) [inline public/js/navigator-js/integration/BackboneRecipe.js:1]: @bassettsj es6 will be in a follow up PR.

@elimak wanted me (and I agree) to start with just the features so as to keep the diff smaller and make the functional changes more easy to distinguish.

File: public/js/navigator-js/integration/StateViewMap.js (hunk 3 above)
Comment (elimak) [inline public/js/navigator-js/integration/StateViewMap.js]: Could you benchmark it in your demo?

File: public/js/navigator-js/integration/ReactRecipe.js (hunk 5 above)
Comment (elimak) [inline public/js/navigator-js/integration/ReactRecipe.js]: Ok, is it something that you can work on after this PR?

Comment (solomon-gumball) [inline public/js/navigator-js/integration/ReactRecipe.js]: Yes!

File: public/js/navigator-js/integration/ViewRecipe.js [hunk 7]
```diff
@@ -38,6 +42,15 @@ this.navigatorjs.integration = this.navigatorjs.integration || {};
 		toView: function(viewClass) {
 			this._viewClass = viewClass;
 
+			// This is where the recipe adds a mixin depending on the type of viewClass
+			// added.  This allows for different handling for different 'types' of
+			// view recipes.
+
+			// TODO: Find a more secure way to determine whether class
+			// extends react component
```
Comment (inv8der) [inline public/js/navigator-js/integration/ViewRecipe.js]: If you import React and Backbone, you should be able to use `viewClass instanceof React.Component` or `viewClass instanceof Backbone.View` to determine what type of view it is.

**Comment by inv8der** [comment]
- [x] Can you fix the indentation in ReactRecipe.js and BackboneRecipe.js?

File: public/js/navigator-js/integration/StateViewMap.js (hunk 3 above)
Comment (elimak) [inline public/js/navigator-js/integration/StateViewMap.js]: > the .remove() method takes elements out of the DOM. Use .remove() when you want to remove the element itself, as well as everything inside it. In addition to the elements themselves, all bound events and jQuery data associated with the elements are removed. To remove the elements without removing data and events, use .detach() instead.

Ran https://jsperf.com/removing-an-element-from-the-screen/3 several times and got an average of 18% slower with adding / removing DOM elements compare to display none/block - Its pretty significant.

File: public/js/navigator-js/integration/ViewRecipe.js [hunk 8]
```diff
@@ -38,6 +42,15 @@ this.navigatorjs.integration = this.navigatorjs.integration || {};
...
+
```
Comment (inv8der) [inline public/js/navigator-js/integration/ViewRecipe.js:51]: You might be able to do something like this. I tested it with a simple class that just extended Backbone.View. I imagine it would work with React.Component as well.

```
Backbone.View.prototype.isPrototypeOf(viewClass.prototype)
```

File: public/js/navigator-js/integration/StateViewMap.js (hunk 3 above)
Comment (gregegan) [inline public/js/navigator-js/integration/StateViewMap.js]: Side comment (i have not been following the discussion too closely) - I understand JSPerf is a good indicator of performance, but keep in mind that is over like hundreds of thousands (or millions) of operations on one function. When we may possibly only really going to be performing this a few times? 18% doesn't seem overly concerning to me. But I do understand every little bit helps.

File: public/js/navigator-js/integration/StateViewMap.js [hunk 9]
```diff
@@ -77,20 +136,52 @@ this.navigatorjs.integration = this.navigatorjs.integration || {};
...
+				var $proxy = $(document.createElement('div'));
+				$container.append($proxy);
+				ReactDOM.render(recipe._viewInstance, $proxy[0]);
+
+				break;
+
+			case 'REACT > REACT':
+				parentRecipe._showChild(recipe);
+
+				while (parentRecipe.getParentRecipe()) {
```
Comment (inv8der) [inline public/js/navigator-js/integration/StateViewMap.js]: Can you explain a little bit what you are doing here? 

What about when you have BACKBONE > REACT > REACT? Is that a valid recipe combination? It seems like this would set parentRecipe to be the BACKBONE recipe. In the BACKBONE > REACT case, you create a proxy to render the React component.

File: public/js/navigator-js/integration/StateViewMap.js [hunk 10]
```diff
@@ -77,20 +136,52 @@ this.navigatorjs.integration = this.navigatorjs.integration || {};
...
+				break;
+
+			case 'BACKBONE > REACT':
+				var $proxy = $(document.createElement('div'));
+				$container.append($proxy);
+				ReactDOM.render(recipe._viewInstance, $proxy[0]);
+
+				break;
+
+			case 'REACT > REACT':
```
Comment (inv8der) [inline public/js/navigator-js/integration/StateViewMap.js]: Can you explain a little bit what you are doing here?

File: public/js/navigator-js/integration/StateViewMap.js [hunk 11]
```diff
@@ -77,20 +136,52 @@ this.navigatorjs.integration = this.navigatorjs.integration || {};
...
+				parentRecipe._showChild(recipe);
+
+				while (parentRecipe.getParentRecipe()) {
+					parentRecipe = parentRecipe.getParentRecipe();
```
Comment (inv8der) [inline public/js/navigator-js/integration/StateViewMap.js]: What about when you have BACKBONE > REACT > REACT? Is that a valid recipe combination? It seems like this would set parentRecipe to be the BACKBONE recipe. In the BACKBONE > REACT case, you create a proxy to render the React component.

Comment (solomon-gumball) [inline public/js/navigator-js/integration/StateViewMap.js]: This is possible. `BACKBONE > REACT > REACT` is already covered here that would be a parent recipe with a `BACKBONE > REACT` combination, and a child with `REACT > REACT`.  The proxy object is added only for calls to substitute in for calls to`transitionIn` and `transitionOut`.

File: public/js/navigator-js/integration/StateViewMap.js (hunk 10 above)
Comment (solomon-gumball) [inline public/js/navigator-js/integration/StateViewMap.js]: So this is a function that handles adding a recipe to a parent recipe.  Previously this would just add one backbone view el to a parent view's el.

The reason for the addition of the switch statement here is that depending on the type (either `backbone` or `react` right now) the way you add the two recipes together changes.

So the process of adding a backbone view to a parent backbone view: `BACKBONE > BACKBONE` is going to be different than adding a react recipe to a backbone recipe `BACKBONE > REACT` or a backbone element to a react parent `REACT > BACKBONE` (this one not implemented yet though).

File: public/js/navigator-js/integration/StateViewMap.js (hunk 11 above)
Comment (inv8der) [inline public/js/navigator-js/integration/StateViewMap.js]: Okay, the reason I asked is because when you have a `BACKBONE > REACT > REACT` situation, it looks like this code will keep traversing up the recipe tree until the parentRecipe is `BACKBONE`, which may be a bug.

File: webpack.config.js (hunk 2 above)
Comment (solomon-gumball) [inline webpack.config.js:23]: I can also fix this in the follow up PR.  Should be an easy fix just a lot of files to change and I didn't want to pollute the diff.

File: karma.conf.js (hunk 4 above)
Comment (solomon-gumball) [inline karma.conf.js]: Removed.

File: public/js/navigator-js/integration/ViewRecipe.js (hunk 8 above)
Comment (solomon-gumball) [inline public/js/navigator-js/integration/ViewRecipe.js:51]: Ended up changing the spacing on StateViewMap to match the indentation of the rest of the repo.

Comment (solomon-gumball) [inline public/js/navigator-js/integration/ViewRecipe.js:51]: Here I ended up splitting into two methods to simplify things.  So it's `toView` and `toComponent` now.  Was going to go down the peer dependency route but testing is an issue.

File: public/js/navigator-js/integration/ViewRecipe.js [hunk 12]
```diff
@@ -1,15 +1,20 @@
+var ReactRecipe = require('./ReactRecipe');
+var BackboneRecipe = require('./BackboneRecipe');
+var ReactTestUtils = require('react-addons-test-utils');
```
Comment (inv8der) [inline public/js/navigator-js/integration/ViewRecipe.js]: I don't see `ReactTestUtils` being used in this file

Comment (solomon-gumball) [inline public/js/navigator-js/integration/ViewRecipe.js]: You are correct sir, that's a leftover I'll remove that guy.

File: public/js/navigator-js/integration/StateViewMap.js [hunk 13]
```diff
@@ -1,111 +1,161 @@
...
+        statesLength = recipeStates.length;
+
+        for (j = 0; j < statesLength; j++) {
+          state = recipeStates[j];
+
+          if (requestedState.contains(state)) {
+
+            this._addViewElementToDOM(recipe);
+            var viewInstance = recipe.getViewInstance();
+            if (Array.isArray(viewInstance.navigatorBehaviors)) {
```
Comment (solomon-gumball) [inline public/js/navigator-js/integration/StateViewMap.js:58]: This line was changed from `_.isArray`

File: public/js/navigator-js/integration/StateViewMap.js [hunk 14]
```diff
@@ -1,111 +1,161 @@
...
+              this._navigator.add(viewInstance, state);
+            }
+          }
+        }
+      }
+    },
+
+    _addRecipeToParent: function _addRecipeToParent(parentRecipe, recipe) {
```
Comment (solomon-gumball) [inline public/js/navigator-js/integration/StateViewMap.js:66]: This is an entirely new function

File: public/js/navigator-js/integration/StateViewMap.js [hunk 15]
```diff
@@ -1,111 +1,161 @@
...
+          );
+          break;
+        default:
+          console.error('Invalid recipe type combination: ' + additionType);
+          break;
+
+      }
+    },
+
+    _addViewElementToDOM: function _addViewElementToDOM(recipe) {
```
Comment (solomon-gumball) [inline public/js/navigator-js/integration/StateViewMap.js:130]: This function was changed in that the actual addition of a recipe to the DOM was split into a seperate function, and this function just verifies that the recipe is initialized and that it needs mounting.

File: public/js/navigator-js/integration/StateViewMap.js [hunk 16]
```diff
@@ -1,111 +1,161 @@
...
+      if (parentRecipe) {
+        // If parent view recipe has not been varructed, initialize
+        // the parent view and add it to the DOM correctly
+
+        if (!parentRecipe.isMounted()) {
+          this._addViewElementToDOM(parentRecipe);
+        }
+      }
+
+      this._addRecipeToParent(parentRecipe, recipe);
```
Comment (solomon-gumball) [inline public/js/navigator-js/integration/StateViewMap.js:155]: This is where the magic happens. :baby:

File: public/js/navigator-js/integration/StateViewMap.js [hunk 17]
```diff
@@ -1,111 +1,161 @@
...
+      for (i = 0; i < length; i++) {
+        recipe.addState(navigatorjs.NavigationState.make(statesOrPaths[i]));
+      }
+
+      this._orderedRecipes.push(recipe);
+
+      return recipe;
+    },
+
+    _handleStateRequested: function _handleStateRequested(e, eventData) {
```
Comment (solomon-gumball) [inline public/js/navigator-js/integration/StateViewMap.js:40]: This and the above functions were not changed.  Sorry for the big diff.

**Comment by elimak** [comment]
- [x] Can you create a proxy as well for updateState?

**Comment by bassettsj** [comment]
:+1:

File: package.json [hunk 18]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "@gigwalk/navigator-js",
   "version": "0.5.0",
```
Comment (gregegan) [inline package.json:3]: can you bump this version #? Then we can merge it in. `0.6.0`

Comment (bassettsj) [inline package.json:3]: This should be done after this PR.
